library/Rules/AbstractComposite.php

```diff
@@ -41,86 +59,61 @@ public function setName(string $name): Validatable
-            $this->appendRule($validator);
+        if ($this->shouldHaveNameOverwritten($rule)) {
+            $rule->setName($this->getName());
         }
 
-        return $this;
-    }
-
-    public function removeRules(): void
-    {
-        $this->rules = [];
-    }
-
-    public function addRules(array $validators)
-    {
-        foreach ($validators as $key => $spec) {
-            if ($spec instanceof Validatable) {
-                $this->appendRule($spec);
-            } elseif (is_numeric($key) && is_array($spec)) {
-                $this->addRules($spec);
-            } elseif (is_array($spec)) {
-                $this->addRule($key, $spec);
-            } else {
-                $this->addRule($spec);
-            }
-        }
+        $this->rules[] = $rule;
 
         return $this;
     }
 
-    public function getRules()
+    /**
+     * Returns all the rules in the stack.
+     *
+     * @return Validatable[]
+     */
+    public function getRules(): array
     {
         return $this->rules;
     }
 
-    public function hasRule($validator)
+    /**
+     * Validate the whole stack of rule.
+     *
+     * @param mixed $input
+     *
+     * @return ValidationException[]
+     */
+    protected function validateRules($input): array
     {
-        if (empty($this->rules)) {
-            return false;
-        }
-
-        if ($validator instanceof Validatable) {
-            return isset($this->rules[spl_object_hash($validator)]);
-        }
-
-        if (is_string($validator)) {
-            foreach ($this->rules as $rule) {
-                if (get_class($rule) == __NAMESPACE__.'\\'.$validator) {
-                    return true;
-                }
-            }
-        }
-
-        return false;
+        return array_filter(
+            array_map(
+                function (Validatable $rule) use ($input): ?ValidationException {
+                    try {
+                        $rule->assert($input);
+                    } catch (ValidationException $exception) {
+                        return $exception;
+                    }
+
+                    return null;
+                },
+                $this->getRules()
+            )
+        );
     }
 
-    protected function appendRule(Validatable $validator): void
+    private function shouldHaveNameOverwritten(Validatable $rule): bool
     {
-        if (!$validator->getName() && $this->getName()) {
-            $validator->setName($this->getName());
-        }
-
-        $this->rules[spl_object_hash($validator)] = $validator;
-    }
-
-    protected function validateRules($input)
-    {
-        $validators = $this->getRules();
-        $exceptions = [];
-        foreach ($validators as $v) {
-            try {
-                $v->assert($input);
-            } catch (ValidationException $e) {
-                $exceptions[] = $e;
-            }
-        }
-
-        return $exceptions;
+        return null !== $this->getName() && null === $rule->getName();
```

Comment:
maybe add a `hasName(): bool` would be better? since name is optional, then we could check for it.

Reply:
That would be handy. In other hand I think the `Validatable` already has too many methods